runtime/src/accounts_db.rs

```diff
@@ -179,6 +179,73 @@ pub enum StoreReclaims {
     Ignore,
 }
 
+/// while combining into ancient append vecs, we need to keep track of the current one that is receiving new data
+/// The pattern for callers is:
+/// 1. this is a mut local
+/// 2. do some version of create/new
+/// 3. use it (slot, append_vec, etc.)
+/// 4. re-create it sometimes
+/// 5. goto 3
+/// If a caller uses it before initializing it, it will be a runtime unwrap() error, similar to an assert.
+/// That condition is an illegal use pattern and is justifiably an assertable condition.
+#[derive(Default)]
+struct CurrentAncientAppendVec {
+    slot_and_append_vec: Option<(Slot, Arc<AccountStorageEntry>)>,
+}
+
+impl CurrentAncientAppendVec {
+    fn new(slot: Slot, append_vec: Arc<AccountStorageEntry>) -> CurrentAncientAppendVec {
+        Self {
+            slot_and_append_vec: Some((slot, append_vec)),
+        }
+    }
+    fn create_ancient_append_vec(&mut self, slot: Slot, db: &AccountsDb) {
+        *self = Self::new(slot, db.create_ancient_append_vec(slot));
+    }
+    fn create_if_necessary(&mut self, slot: Slot, db: &AccountsDb) {
+        if self.slot_and_append_vec.is_none() {
+            self.create_ancient_append_vec(slot, db);
+        }
+    }
+
+    /// note this requires that 'slot_and_append_vec' is Some
+    fn slot(&self) -> Slot {
+        self.slot_and_append_vec.as_ref().unwrap().0
+    }
+
+    /// note this requires that 'slot_and_append_vec' is Some
+    fn append_vec(&self) -> &Arc<AccountStorageEntry> {
+        &self.slot_and_append_vec.as_ref().unwrap().1
+    }
+
+    /// note this requires that 'slot_and_append_vec' is Some
+    fn append_vec_id(&self) -> AppendVecId {
+        self.append_vec().append_vec_id()
+    }
+
+    /// helper function to cleanup call to 'store_accounts_frozen'
+    fn store_ancient_accounts(
```

Comment:
Feels like this should belong to AccountsDb rather than CurrentAncientAppendVec to me. A sort of circular dependence going on here. We are generally passing CurrentAncientAppendVec around to AccountsDb fns, then we reverse it for this function.

Reply:
`create_ancient_append_vec` follows the same pattern. There is high coupling between these structs. Trying to group related functionality together.